File: feature-venue-detail/src/main/java/com/smedialink/feature_venue_detail/state/VenueViewState.java
package com.smedialink.feature_venue_detail.state;

import home.oleg.placesnearme.core_presentation.base.ErrorEvent;
import home.oleg.placesnearme.core_presentation.viewdata.VenueViewData;

public class VenueViewState {
    private final boolean isLoading;
    private final ErrorEvent errorEvent;
    private final VenueViewData venueViewData;

    private VenueViewState(boolean isLoading, ErrorEvent errorEvent, VenueViewData venueViewData) {
        this.isLoading = isLoading;
        this.errorEvent = errorEvent;
        this.venueViewData = venueViewData;
    }

    public static VenueViewState loading() {
        return new VenueViewState(true, null, null);
    }

    public static VenueViewState error(ErrorEvent errorEvent) {
        return new VenueViewState(false, errorEvent, null);
    }

    public static VenueViewState success(VenueViewData venueViewData) {
        return new VenueViewState(false, null, venueViewData);
    }

    public boolean isLoading() {
        return isLoading;
    }

    public ErrorEvent getErrorEvent() {
        return errorEvent;
    }

    public VenueViewData getVenueViewData() {
        return venueViewData;
    }
}

File: feature-venue-detail/src/main/java/home/oleg/placesnearme/feature_venue_detail/presentation/VenueViewState.kt
package home.oleg.placesnearme.feature_venue_detail.presentation

import com.olegsheliakin.statebinder.State
import com.smedialink.common.base.ErrorEvent
import com.smedialink.common.base.MessageEvent
import home.oleg.placesnearme.corepresentation.viewdata.PlaceViewData

data class PlaceViewState(
        val isLoading: Boolean,
        val error: ErrorEvent?,
        val place: PlaceViewData?,
        val message: MessageEvent?
) : State {

    companion object {
        fun initial() = PlaceViewState(
                isLoading = false,
                error = null,
                place = null,
                message = null)
    }
}

sealed class VenueViewState {
    object Initial : VenueViewState()
    object Loading : VenueViewState()
    data class Success(val venue: PlaceViewData) : VenueViewState()
    data class Error(val errorEvent: ErrorEvent) : VenueViewState()
}
